BUG: Ignore differences in NAN for computing ULP differences

r-devulap opened the pull request:
Fixes https://github.com/numpy/numpy/issues/15593

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

mattip commented on numpy/core/tests/test_umath_accuracy.py:
```diff
@@ -50,5 +50,7 @@ def test_validate_transcendentals(self):
                         inval = inval[perm]
                         outval = outval[perm]
+                        inval[np.argwhere(np.isnan(inval))] = np.nan
+                        outval[np.argwhere(np.isnan(inval))] = np.nan
```
a comment here would be helpful

seberg replied:
Can we fix this within the `assert_array_max_ulp` function? That function should be consistent in either rejecting or accepting NaN I think.

mattip replied:
`assert_array_max_ulp`  indeed does not document its NaN behaviour, so I guess we can document and fix it there.

mattip replied:
also it is a bit annoying that `assert_array_max_ulp` does not show the offending output like `assert_array_compare` but only the maximum ULP discovered.

r-devulap replied:
makes sense, I can update the `assert_array_max_ulp ` function instead. Where would you like me to add the document?

mattip replied:
the docstring of the function itself does not indicate any special handling of NaNs. If we change that, we should document it

mattip replied:
in the docstring

Commit:
BUG: Ignore differences in NAN representation for computing ULP differences

seberg commented:
Thanks, do you think it might make sense to push it even further down to the `integer_representation` function? Since the "signed magnitude" representation is not just a view as far as I understand it, that might make sense?

r-devulap commented:
not so sure about that, it is the max ULP diff function that doesn't want to differentiate between NAN's. The `integer_representation`  function simply returns the binary representation of a floating point and I think that should stay true to its intent for NAN's as well.

r-devulap commented:
`x[np.argwhere(np.isnan(x))] = np.nan` fails with `E       IndexError: too many indices for array` for a single element array like `x = np.array(1.2) ` Any suggestions on how to elegantly fix it without special casing it?

seberg commented:
@r-devulap any reasons you are using `np.argwhere` at all, just remove it? Not sure I agree with the interpretation of `integer_representation`, but it also does not really matter much right now.

r-devulap commented:
> @r-devulap any reasons you are using `np.argwhere` at all, just remove it? 

Thanks :)

Commit:
TST: remove large numbers from sin/cos validation test suite

sine and cosine of large numbers are always computed via the standard C
library and the results vary depending on the platform.

seberg commented:
@r-devulap looks good to me, would you be up for adding a test as well, since I think the ulp thing is arguably a bug fix in our testing utils, and also testing utils need testing :).

r-devulap commented:
Sure, what do you have in mind? ensure ulp error between two different representations of NAN = 0?

Commit:
TST: Add test to ensure assert_array_max_ulp ignores ULP differences betwen NAN's

r-devulap commented:
@seberg let me know if that suffices.

seberg commented:
Test would be nicer if you move it into `numpy/testing/tests/test_utils.py`. Otherwise looks good to me, I will let Matti sign off on the line removal changes and merge. Thanks!

mattip commented:
I think there is more to do with the umath_accuracy test but this solves the current issues. xref gh-15605 for some possible improvements

mattip commented:
Thanks @r-devulap for the quick response
